### DISCUSSION TITLE
Solving a PDE problem considering the nonlinear diffusion coefficient of Richards equation

### DISCUSSION THREAD

**Discussion by essicolo** [usnistgov/fipy#965] [General]
Hello Fipy's community,

I'm working on a PDE problem, which was recently reformulated by Dr. Guyer on [Stackoverflow](https://stackoverflow.com/questions/77187818/solving-a-pde-problem-considering-the-nonlinear-diffusion-coefficient-of-richard). The problem is represented as:

$$
\frac{\partial \theta}{\partial t} + \nabla \cdot ( k(u) \nabla (u+z) ) - S = 0
$$

↳ I've edited this equation after @guyer's [comment](https://github.com/usnistgov/fipy/discussions/965#discussioncomment-7256925).

$$
\frac{\partial \theta}{\partial t} + \nabla \cdot ( k(u) \nabla u ) + \nabla \cdot ( k(u) \nabla z ) - S = 0
$$

$$
\frac{\partial \theta}{\partial t} + \nabla \cdot ( k(u) \nabla u ) + \nabla \cdot ( k(u) \hat{k} ) - S = 0
$$

$$
\frac{\partial \theta(u)}{\partial u} \times \frac{\partial u}{\partial t}  + \nabla \cdot ( k(u) \nabla u ) + \nabla \cdot ( k(u) \hat{k} ) - S = 0
$$

Without the sink/source term and in steady-stade, we have:

$$
\nabla \cdot ( k(u) \nabla u ) + \nabla \cdot ( k(u) \hat{k} ) = 0
$$

Where:
- $\hat{k}$ is a unit vector in the $z$ direction (elevation).
- $\theta(u)$ is a non-linear function that is ruled out in steady state
- $k(u)$ is a non-linear function defined as:

$$
k(u) = k_{sat} \times (1 - (-a ~ u)^{n-1} \times (1 + (-a ~ u)^n)^{-m})^2 \times (1 + (-a ~ u)^n)^{-m ~ l}
$$

With $k_{sat}$, $a$, $n$, $m$, and $l$ being material parameters.

**Issue**. After implementing Dr. Guyer's suggestions, my solution doesn't align with two analytical benchmarks. The expected pressure head should start at zero (the Dirichlet BC), then decrease and stabilize to a minimum value at a certain elevation from the water table. However, the FiPY solution appears influenced by the column height (`Lz`). The initial guess also heavily impacts the results. For instance, with an initial guess of 0.0, all `u` values (pressure head) remain at zero. Different solvers yield similar outcomes.

**Request**. I'm seeking suggestions to adjust my code in a way it fits approximately to my benchmarks. You can review it [here](https://gist.github.com/essicolo/57cb63cc6fe5c3f93da1a91c9b8a8a8f).

Thank you for your assistance!

Best regards,

Essi

**Comment by guyer** [comment]
> my solution doesn't align with two analytical benchmarks. The expected pressure head should start at zero (the Dirichlet BC), then decrease and stabilize to a minimum value at a certain elevation from the water table. However, the FiPY solution appears influenced by the column height (`Lz`). 

Why wouldn't it be sensitive to the column height? What in the math would cause it to stabilize at some arbitrary elevation? I'm afraid $k(u)$ is too complicated for me to have any intuition about its value or about what effect it might have. I see that $k(u)$ has an inflection at intermediate depth, but I still don't know why that should mean that $u$  should saturate.

> For instance, with an initial guess of 0.0, all `u` values (pressure head) remain at zero. Different solvers yield similar outcomes.

$u = 0$ is a solution to the steady state equation.

**Reply by essicolo** [reply]
> Why wouldn't it be sensitive to the column height? What in the math would cause it to stabilize at some arbitrary elevation?

I expect that in steady state the total height of a column in shouldn't change the profile of the capillarity tension (suction, or negative pressure) above the water table. Physically, where water flows well above the water table, the gradient  $i$ is 1. The gradient is the total head discharge, such as

$$i = \frac{\partial h}{\partial z} = \frac{\partial (u+z)}{\partial z}$$

At some point near the water table, the gradient quickly becomes 0. When the gradient is 1, pressure head is steady, and when it's zero, pressure equals minus the elevation. Mathematically, it's a consequence of Dary's law known as the [Kisch (1959)](https://doi.org/10.1680/geot.1959.9.1.9) formulation.

$$ z_j = \int_{0}^{u_j} \frac{du}{1 - \frac{q}{k(u)}} $$

The inflection in the pressure profile can also be predicted on the hydraulic conductivity curve, $k(u)$. Parameters $k_{sat}$, $a$, $n$, $m$, and $l$ drive the shape of this curve.

![linear-method](https://github.com/usnistgov/fipy/assets/5118227/9bfd049c-1dbd-4dee-b2f3-dfb2c65639d0)

I updated the code according to your suggestions in the comment section of the gist file. But I found out that there is almost an exact match between the Kisch (1959) equation and Fipy only where `Lz=6.69`. I can't figure out why though.

Thanks for your time!

Essi

**Reply by guyer** [reply]
You're using words to describe this and I only have a superficial understanding of what those words mean. Moreover, I don't have access to the reference you provided.

*In the math*, what is "water table"? What is $z_j$? What is the relationship between the symbols you have used and the words you have used to label the graphs?

**Reply by essicolo** [reply]
> *In the math*, what is "water table"?

Oh I see. In the math, the water table is everywhere in the domain where $u=0$. In my problem, $u=0$ is a Dirichlet boundary condition in `u.constrain(0., mesh.facesLeft)`. 

> What is $z_j$?

$z_j$ is the $j^{th}$ elevation from a reference. It can be computed with the Kisch form of Darcy's law from a vector of pressures $u$ transformed in a vector of hydraulic conductivities $k(u)$, and for a given infiltration flux $q$.

> What is the relationship between the symbols you have used and the words you have used to label the graphs?

- $u$ is the pressure, the cell variable to be solved. It can be expressed in "meters of water"
- $z$ in the elevation, or the position from left to right in a 1D mesh.
- $h$ is total head, $h=u+z$
- $i$ is the gratient, $i_x = \frac{\partial h}{\partial x}$, where $x$ is any dimension in space
- $q=1E-6$ in the first graph is the right (or top) Neumann boundary condition. It can be used to compute $u_q = 1.73~m$, in my case study

In theory, the pressure in a soil column ($u$) with $u=0$ at its base and a flow $q$ at its top should decrease linearly with elevation then level off at $u_q$ a elevation $z_q = u_q$.

Have expressed the problem correctly, in understandable math? Does my gist code reflect these concepts?

**Reply by guyer** [reply]
Better, thank you. Where is $q$ in your PDE? You use it to calculate the initial condition and then never again as far as I can tell.

Presently, at the right boundary, you have FiPy's natural boundary condition of zero flux (this is the natural boundary condition of the finite volume method). Note that, for your problem, flux is $-k(u)\nabla(u + z)$. Zero flux does *not* mean that $\hat{n}\cdot\nabla u$ is equal zero.

**note:** I think you have an error in your first equation at top. I think it should be

$$
\frac{\partial \theta}{\partial t} + \nabla \cdot ( k(u) \nabla (u+z) ) - S = 0
$$

**Reply by essicolo** [reply]
In fact, maybe the whole problem should be expressed in terms of total head $h$ ($h=u+z$)? This will make the code easier to express since, as you wrote, the flux is related to the gradient of $h$, not $u$ — it took me so much time to figure that out...

**Action**. I rearranged the [code](https://gist.github.com/essicolo/57cb63cc6fe5c3f93da1a91c9b8a8a8f) to solve for `h` instead. I'm wondering if, once expressed on $h$, is the `+ (k.faceValue * [[1]]).divergence` term is still needed?

**Results**. It seems that `eq.sweep(var=h, solver=solver)` didn't alter the initial guess and all $h$ values are zeros.

**Reply by guyer** [reply]
Yes, solving for $h$ is probably better, in which case, no, you don't need `(k.faceValue * [[1]]).divergence`.

I'm not getting $h = 0$ from revision 13.

Please express your boundary conditions mathematically. What exactly is $q$?

**Reply by essicolo** [reply]
EDITED: You are right, I'm not getting $h=0$ either. Moreover, there was a sign error, in `if np.max(np.abs(h.value - h.old.value)) > tolerance`, which obviously should have been `if np.max(np.abs(h.value - h.old.value)) < tolerance`. Now `h` values are `Inf`.

> What exactly is q?

I'll do my best, mathematically explained. $q$ in the context of my code is a flux going inward the domain, such as

$$
q = - \hat{n} \cdot k(u) \nabla h
$$

where $u$ is a cell variable such as $u = h-z$, and $z$ is the dimension of the domain representing elevation.

The Neumann boundary condition expressed as $\nabla h$ is thus

$$
\nabla h = - \frac{q}{\hat{n} \cdot k(u) }
$$

which I would express in my FiPY problem as `h.faceGrad.constrain(- q/k.faceValue, where=mesh.facesRight)`.

Thanks again for your time!

**Reply by guyer** [reply]
OK, I assumed $q$ was something like that, but wanted to be sure. Just to confirm, it's definitely $q=-\hat{n}\cdot k(u)\nabla h$ and not $q=-\hat{n}\cdot k(u)\nabla u$?

Regardless, I've tried a couple of different formulations (it's actually more natural for finite volume methods to constrain the flux than the gradient[*]), but all cycle between a few (wrong) solutions. This behavior only reinforces two prejudices I've developed over the years: 1D formulations are always wrong and steady-state solutions are nonsense.

If I institute relaxation (artificial time evolution), I am able to obtain good agreement with your Kisch solution.
![relaxation solution](https://github.com/usnistgov/fipy/assets/1830725/1411381e-c111-4d8d-9acc-f7057bf96dc1).

Specifically,
```diff
diff --git a/richards.py b/richards.py
index ee8ef7a..669cdb3 100644
--- a/richards.py
+++ b/richards.py
@@ -2,6 +2,8 @@ import numpy as np
 from scipy.optimize import fsolve
 import matplotlib.pyplot as plt
 import fipy as fp
+from petsc4py import PETSc
+
 PETSc.Options().setValue("snes_type", "newtonls")
 
 def vg_k(u, k_sat, a, n, m, l=0.5):
@@ -33,24 +35,23 @@ def richards_fem(q, k_sat, a, n, m, l=0.5, Lz=10, nz=100, max_iters=100, toleran
     # Variables
     z = fp.CellVariable(name="elevation", mesh=mesh, value=mesh.cellCenters[0])
     h = fp.CellVariable(name="total head", mesh=mesh, value=0.0, hasOld=True)
-    u = fp.CellVariable(name="matrix pressure", mesh=mesh, value=h.value-z.value, hasOld=True)    
-    k = fp.CellVariable(mesh=mesh, value=vg_k(u.value, k_sat, a, n, m, l))
+    u = h - z
+    u.name = "matrix pressure"
+    k = vg_k(u.faceValue, k_sat, a, n, m, l)
 
     # Boundary conditions
     h.constrain(0.0, mesh.facesLeft) # Dirichlet at bottom
-    h.faceGrad.constrain(- q/k.faceValue, where=mesh.facesRight)  # Neumann at top
+    h.faceGrad.constrain(q/k, where=mesh.facesRight)  # Neumann at top
 
     # Equation
-    eq = (fp.DiffusionTerm(coeff=k, var=h) == 0)
+    eq = (fp.TransientTerm(var=h) == fp.DiffusionTerm(coeff=k, var=h))
 
     # Solve with Picard iterations
     solver = fp.DefaultSolver(tolerance=1e-10, iterations=5000)
     for iter in range(max_iters):
-        u.value = h.value - z.value
-        k.value = vg_k(u.value, k_sat, a, n, m, l)
-        eq.sweep(var=h, solver=solver)
-        if np.max(np.abs(h.value - h.old.value)) > tolerance:
-            break
+        h.updateOld()
+        for sweep in range(3):
+            res = eq.sweep(dt=10000., var=h, solver=solver)
 
     return u.value, z.value
 
@@ -73,11 +74,12 @@ z_kisch = kisch(q=rain, u=u_kisch, u_q=u_q, k_sat=k_sat, a=a, n=n, m=m, l=l, z_b
 
 u_fem, z_fem = richards_fem(
     q=rain, k_sat=k_sat, a=a, n=n, m=m, l=l,
-    Lz=10, nz=100, max_iters=1000, tolerance=1e-8
+    Lz=10, nz=1000, max_iters=10000, tolerance=1e-8
 )
```
Notes:
- I have rewritten `u` and `k` to take advantage of FiPy's lazy evaluation. There is no need to manually set them after every solution.
- I had to flip the sign of your Neumann boundary condition.
- Time steps above about 10000 were unstable.
- Your `if np.max(np.abs(h.value - h.old.value)) > tolerance:` is backwards, but also isn't really helpful with relaxation and sweeping.

[*] To constrain flux, rather than gradient, make these additional changes:
```diff
diff --git a/richards.py b/richards.py
index 669cdb3..ab1d307 100644
--- a/richards.py
+++ b/richards.py
@@ -41,10 +41,11 @@ def richards_fem(q, k_sat, a, n, m, l=0.5, Lz=10, nz=100, max_iters=100, toleran
 
     # Boundary conditions
     h.constrain(0.0, mesh.facesLeft) # Dirichlet at bottom
-    h.faceGrad.constrain(q/k, where=mesh.facesRight)  # Neumann at top
 
     # Equation
-    eq = (fp.TransientTerm(var=h) == fp.DiffusionTerm(coeff=k, var=h))
+    eq = (fp.TransientTerm(var=h)
+          == fp.DiffusionTerm(coeff=k, var=h)
+          + (q * mesh.faceNormals * mesh.facesRight).divergence)
 
     # Solve with Picard iterations
     solver = fp.DefaultSolver(tolerance=1e-10, iterations=5000)
```

**Comment by essicolo** [comment]
This look promising! However, I couldn't test the results since the patch returned an index error

`IndexError: diffusion coefficient tensor is not an appropriate shape for this mesh`

Have you changed the mesh to higher dimensions?

**Reply by guyer** [reply]
I made only the changes I showed. Update your gist and I'll take a look.

**Reply by essicolo** [reply]
Ok thanks. Gist updated!

**Reply by guyer** [reply]
You can't use `np.where` that way. NumPy doesn't know anything about FiPy's lazy mesh variables.

You can do this:
```diff
diff --git a/richards.py b/richards.py
index 0ea2b1c..d0beab1 100644
--- a/richards.py
+++ b/richards.py
@@ -7,11 +7,8 @@ from petsc4py import PETSc
 PETSc.Options().setValue("snes_type", "newtonls")
 
 def vg_k(u, k_sat, a, n, m, l=0.5):
-    k = np.where(
-        u>=0,
-        k_sat,
-        k_sat * (1 - (-a * u)**(n-1) * (1 + (-a * u)**n)**(-m))**2 * (1 + (-a * u)**n)**(-m*l)
-    )
+    k = k_sat * ((u >= 0) * 1.
+                 + (u < 0) * (1 - (-a * u)**(n-1) * (1 + (-a * u)**n)**(-m))**2 * (1 + (-a * u)**n)**(-m*l))
     return k
 
 def find_u_for_given_k(target_k, k_sat, a, n, m, l=0.5, u_guess=-1.0):
```

**Reply by essicolo** [reply]
Hello Dr Gruyer,

I'm pleased to inform you that the simulation was a success! I've made some modifications to the code, which you can view in this [gist](https://gist.github.com/essicolo/57cb63cc6fe5c3f93da1a91c9b8a8a8f). Remarkably, after just a few iterations, the results align closely with the Kisch profile.

In the coming days, I plan to further enhance the code by exploring higher dimensions, incorporating different materials, and introducing more intricate boundary conditions. The foundation you've provided has opened up numerous possibilities, and I'm genuinely enthusiastic about the future potential of this project.

Thank you once again for your invaluable assistance.

--
Essi
